File: vtk-rs/libvtkrs/src/vtk_interactor_style_trackball_camera.cpp
#include "cxx.h"
#include "vtk_interactor_style_trackball_camera.h"
#include "vtk_interactor_style_trackball_camera.rs.h"

#include <vtkInteractorStyleTrackballCamera.h>

vtkInteractorStyleTrackballCamera* interactor_style_trackball_camera_new() {
    vtkInteractorStyleTrackballCamera* obj = vtkInteractorStyleTrackballCamera::New();
    if (!obj) {
        throw std::runtime_error("Failed to create vtkInteractorStyleTrackballCamera");
    }
    return obj;
}

void interactor_style_trackball_camera_delete(vtkInteractorStyleTrackballCamera& style) {
    style.Delete();
}

File: vtk-rs/src/vtk_interactor_style_trackball_camera.rs
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("vtk_interactor_style_trackball_camera.h");

        type vtkInteractorStyleTrackballCamera;

        fn interactor_style_trackball_camera_new() -> *mut vtkInteractorStyleTrackballCamera;
        fn interactor_style_trackball_camera_delete(
            style: Pin<&mut vtkInteractorStyleTrackballCamera>
        );
    }
}

crate::define_object!(
    "https://vtk.org/doc/nightly/html/classvtkInteractorStyleTrackballCamera.html",
    @name InteractorStyleTrackballCamera, ffi::vtkInteractorStyleTrackballCamera,
    @new ffi::interactor_style_trackball_camera_new,
    @delete ffi::interactor_style_trackball_camera_delete
);

pub(crate) mod private {
    pub trait Sealed {}
}

/// [`vtkInteractorStyleTrackballCamera`](https://vtk.org/doc/nightly/html/classvtkInteractorStyleTrackballCamera.html)
///
/// Interactive manipulation of the camera with trackball-style controls:
/// - Left mouse button: Rotate the camera around the scene
/// - Right mouse button: Zoom in/out
/// - Middle mouse button: Pan the camera
/// - Scroll wheel: Zoom in/out
#[allow(non_camel_case_types)]
pub trait vtkInteractorStyleTrackballCamera: private::Sealed {}
